docs/adr/0004-before-action-variables.md

```diff
@@ -0,0 +1,91 @@
+# 4. Before action variables
+
+Date: 2020-08-11
+
+## Status
+
+Accepted
+
+## Context
+
+This code base currently has a lot of `before_action` in controllers. For
+example:
+
+- https://github.com/DFE-Digital/teacher-training-api/blob/6e18d28cafe3d4595a15eb6670cc6950ddf692b8/app/controllers/api/v2/courses_controller.rb#L6
+
+These `before_action` calls typically load data into an instance variable
+which are then used later.
+
+However this can become unwieldy when the controller has multiple actions.
+Some actions require one set of variables whilst another action may not and
+there are resulting overlaps between the variables in use.
+
+Therefore to manage the actions will typically have `only` or `except` options
+set to control these conditions.
+
+Therefore this can cause confusion and cognitive overload when trying to
+comprehend what variables are actually being loaded and used concerning a
+specific action a developer may be working on.
+
+Sometimes because it is not known if these varaibles are needed or not or due
+to a mistake it is possible for expensive calls to made to load data that may
+not even be needed for the action.
+
+## Options
+
+### 1. Do nothing
+
+The first options is to simply not address this issue and carry on with the
+currect pattern.
+
+#### Pros
+
+- No work to change existing code.
+
+#### Cons
+
+- Persistence of overhead of understanding variables loaded in controllers.
+
+### 1. Use lazy loaded memoised private methods
```

Comment:
`2.` ?

Reply:
whoops, yea i did fix this but forgot to push